sw/device/lib/dif/dif_usbdev.h

```diff
@@ -0,0 +1,759 @@
+
+/**
+ * Enable or disable reception of OUT packets for an endpoint.
+ *
+ * @param usbdev A USB device.
+ * @param endpoint An endpoint.
+ * @param new_state New OUT packet reception state.
+ * @return `kDifUsbdevResultInvalidArgument` if `usbdev` is `NULL`,
+ * `kDifUsbdevResultGeneralError` if `new_state` or `endpoint` is invalid,
+ * `kDifUsbdevResultOK` otherwise.
+ */
+dif_usbdev_result_t dif_usbdev_endpoint_out_enable(
+    dif_usbdev_t *usbdev, uint32_t endpoint, dif_usbdev_toggle_t new_state);
+
+/**
+ * Enable or disable STALL for an endpoint.
+ *
+ * @param usbdev A USB device.
+ * @param endpoint An endpoint.
+ * @param new_state New STALL state.
+ * @return `kDifUsbdevResultInvalidArgument` if `usbdev` is `NULL`,
+ * `kDifUsbdevResultGeneralError` if `new_state` or `endpoint` is invalid,
+ * `kDifUsbdevResultOK` otherwise.
+ */
+dif_usbdev_result_t dif_usbdev_endpoint_stall_enable(
+    dif_usbdev_t *usbdev, uint32_t endpoint, dif_usbdev_toggle_t new_state);
+
+/**
+ * Get STALL state of an endpoint.
+ *
+ * @param usbdev A USB device.
+ * @param endpoint An endpoint.
+ * @param state Current STALL state.
+ * @return `kDifUsbdevResultInvalidArgument` if `usbdev` or `state` is `NULL`,
+ * `kDifUsbdevResultGeneralError` if `endpoint` is invalid,
+ * `kDifUsbdevResultOK` otherwise.
+ */
+dif_usbdev_result_t dif_usbdev_endpoint_stall_get(dif_usbdev_t *usbdev,
+                                                  uint32_t endpoint,
+                                                  bool *state);
+
+/**
+ * Enable or disable isochronous mode for an endpoint.
+ *
+ * Isochronous endpoints transfer data periodically. Since isochronous transfers
+ * do not have a handshaking stage, isochronous endpoints cannot report errors
+ * or STALL conditions.
+ *
+ * @param usbdev A USB device.
+ * @param endpoint An endpoint.
+ * @param new_state New isochronous state.
+ * @return `kDifUsbdevResultInvalidArgument` if `usbdev` is `NULL`,
+ * `kDifUsbdevResultGeneralError` if `endpoint` or `new_state` is invalid,
+ * `kDifUsbdevResultOK` otherwise.
+ */
+dif_usbdev_result_t dif_usbdev_endpoint_iso_enable(
+    dif_usbdev_t *usbdev, uint32_t endpoint, dif_usbdev_toggle_t new_state);
+
+/**
+ * Enable the USB interface of a USB device.
+ *
+ * Calling this function causes the USB device to assert the full-speed pull-up
+ * signal to indicate its presence to the host.
+ *
+ * @param usbdev A USB device.
+ * @param new_state New interface state.
+ * @return `kDifUsbdevResultInvalidArgument` if `usbdev` is `NULL`,
+ * `kDifUsbdevResultGeneralError` if `new_state` is invalid,
+ * `kDifUsbdevResultOK` otherwise.
+ */
+dif_usbdev_result_t dif_usbdev_interface_enable(dif_usbdev_t *usbdev,
+                                                dif_usbdev_toggle_t new_state);
+
+/**
+ * Information about a received packet.
+ */
+typedef struct dif_usbdev_rx_packet_info {
+  uint32_t endpoint;
+  uint32_t len;
+  bool is_setup;
+} dif_usbdev_rx_packet_info_t;
```

Comment:
I would suggest that, in structs like this one that the user is intended to inspect, the fields probably want comments, as well.